Sprint 2/step 1

sadova-nadezhda opened the pull request.

sadova-nadezhda committed:
create store

sadova-nadezhda committed:
Update slice.ts

sadova-nadezhda committed:
update slices

ingredients slice
burger-constructor slice

sadova-nadezhda committed:
Revert "update slices"

This reverts commit 039ebf9aa655b8c2cf7c9aed025438e164e0bbd9.

sadova-nadezhda committed:
update store

sadova-nadezhda committed:
update BurgerIngredients

Получение списка ингредиентов от API. Используется в компоненте BurgerIngredients

sadova-nadezhda committed:
update BurgerIngredients

обновила setCurrentIngredient, clearCurrentIngredient для модалки

sadova-nadezhda committed:
update IngredientDetails

Добавление данных о просматриваемом в модальном окне IngredientDetails ингредиенте.

sadova-nadezhda committed:
update BurgerConstructor

Хранение списка ингредиентов для конструктора бургера. Используется в компоненте BurgerConstructor

sadova-nadezhda committed:
Update OrderDetails.tsx

Получение и обновление номера заказа в модальном окне OrderDetails.

sadova-nadezhda committed:
update tabs

Доработка интерфейса навигации по ингредиентам

sadova-nadezhda committed:
update project

fixed types for store
fixed types BurgerIngredients

sadova-nadezhda committed:
refactor

sadova-nadezhda committed:
Update main.tsx

sadova-nadezhda committed:
create dnd

Реализация перетаскивания ингредиентов

sadova-nadezhda committed:
update project

- Подправила перетаскивания ингредиентов
- добавила подсчёт итоговой стоимости бургера
- Проверка типизации

sadova-nadezhda committed:
Update slice.ts

added updateIngredientOrder

sadova-nadezhda committed:
update OrderDetails

-  Возможность создавать заказ

sadova-nadezhda committed:
update BurgerConstructor

-Вложенная сортировка ингредиентов в BurgerConstructor

sadova-nadezhda committed:
Update ConstructorCard.tsx

sadova-nadezhda committed:
update project

sadova-nadezhda committed:
Update slice.ts

mikebars1995 commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -1,64 +1,140 @@
+            ) : (
+              mainIngredients.map((ingredient, index) => (
+                <ConstructorCard
+                  key={ingredient._id}
```
**Нужно исправить**

при добавлении ингредиента в конструктор ему нужно добавлять `уникальный uuid`, чтобы использовать его для key, иначе некорректно работает DND.
Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

А если используете `createSlice`, то это делается в `prepare`

```ts

prepare: (ingredient: TIngredient) => {
        return { payload: { ...ingredient, uniqueId: uuid4() } };
      }


```

Подробнее про `prepare` тут https://redux-toolkit.js.org/api/createSlice#customizing-generated-action-creators

mikebars1995 commented on src/services/order/actions.ts:
```diff
@@ -0,0 +1,23 @@
+  'order/fetchOrder',
+  async ({ ingredients }, { rejectWithValue }) => {
+    try {
+      const response = await fetch('https://norma.nomoreparties.space/api/orders', {
```
**Нужно исправить**


`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

mikebars1995 commented on src/services/order/actions.ts:
```diff
@@ -0,0 +1,23 @@
+        body: JSON.stringify({ ingredients }),
+      });
+
+      if (!response.ok) {
+        throw new Error('Ошибка при оформлении заказа');
+      }
```
**Нужно исправить**



У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)


Работа проделана огромная






Что нужно исправить:


* при добавлении ингредиента в конструктор ему нужно добавлять `уникальный uuid`, чтобы использовать его для key, иначе некорректно работает DND.
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты. 
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера





Что можно улучшить:

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

sadova-nadezhda committed:
update projects

Исправила замечания ревьюера
- добавила уникальный uuid
- вынесла константу BASE_URL
- создала отдельную функцию checkResponse
- создала отдельную функцию request
- очищаю конструктор после успешного получения номера

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
